### ISSUE TITLE
Proposing to replace default invalid state for form elements by indeterminate

### ISSUE THREAD

**Issue by notabene** [w3c/html#1073]
Hi all,

(Please forgive me in advance if some terms are approximate, and feel free to correct me if needed for more relevant naming of states, pseudo-states, pseudo-selectors etc.; I won't be offended the least.)

We (@notabene and @Lausselloic) have a problem with the browser validation occurring on form elements: by default, if a field is `required`, its `:invalid` pseudo-state is triggered as soon as the page is loaded.

The problem we have is that if we style elements based on the `:invalid` pseudo-selector, fields are shown as invalid even though a user has not yet had the chance to interact with it.

If I read correctly, [the spec in its current state for required](https://w3c.github.io/html/sec-forms.html#element-attrdef-input-required) does not specify that a field is invalid as of page load. It just says that it is boolean, either true or false. Understandably it's false by default since it's required but was not filled (or checked if it's radio checkboxes). We think that a third **indeterminate** state should be added, and the browser should, for instance, consider the field as invalid **only when one tabs/clicks out of it** (AKA when the blur event is triggered) or when one submits the form.

This way a field would not default to being invalid as long as one has not yet interacted with it or its containing form.

By the way, we have the same problem, although a bit less prominent, with patterns: if a field requires a pattern, its `:invalid` pseudo-state is triggered as soon as the first character is typed. Could the following option be considered: when field is blurred, compare the field value to the pattern and only then trigger its `:invalid` state.

**Comment by notabene** [comment]
(Also EDITORS, feel free to edit this issue's title if it's too long to your taste) 😉 [wink emoji]

**Comment by emmanuelclement** [comment]
Same here.

**Comment by ffoodd** [comment]
FWIW, `:indeterminate` pseudo-class already exists for checkboxes, radios and progress ([latest 
HTML5 W3C recommendation](https://www.w3.org/TR/html5/disabled-elements.html#selector-indeterminate)).

**Comment by notabene** [comment]
Also, it seems that the problem has been identified too in the [CSS Selector Level 4 draft](https://drafts.csswg.org/selectors-4/#user-pseudos):

> The `:user-invalid` pseudo-class represents an element with incorrect input, but only after the user has significantly interacted with it.

**Comment by ffoodd** [comment]
FWIW, if your input has a placeholder you might deal with the issue by using `:placeholder-shown`.

If placeholder is shown, then input hasn't been focused. However it require to use a placeholder (but as far as I tested this, it works even if placeholder is empty).

[CodePen it or it didn't happen](https://codepen.io/ffoodd/pen/PEzoYO)

**Comment by siusin** [comment]
Thanks @notabene .

We're closing this issue on the W3C HTML specification because the W3C and WHATWG are now [working together on HTML](https://www.w3.org/blog/2019/05/w3c-and-whatwg-to-work-together-to-advance-the-open-web-platform/), and all issues are being discussed on the [WHATWG repository](https://github.com/whatwg/html).

If you filed this issue and you still think it is relevant, please open a [new issue](https://github.com/whatwg/html/issues) on the WHATWG repository and reference this issue (if there is useful information here). Before you open a new issue, please check for existing issues on the WHATWG repository to avoid duplication.

If you have questions about this, please open an issue on the [W3C HTML WG repository](https://github.com/w3c/htmlwg/issues/new) or send an email to public-html@w3.org.
